app/employee-app/backend/modules/database/db_functions.bal

```diff
@@ -0,0 +1,438 @@
+
+            if designations is error {
+                string errorMsg = string `An error occurred when retrieving designations data of ${
+                    careerFunction.careerFunction
+                } !`;
+                log:printError(errorMsg, designations);
+                return error(errorMsg);
+            }
+            careerFunctions.push({
+                id: careerFunction.id,
+                careerFunction: careerFunction.careerFunction,
+                designations
+            });
+        };
+    if iterateError is sql:Error {
+        string errorMsg = string `An error occurred when retrieving career function data!`;
+        log:printError(errorMsg, iterateError);
+        return error(errorMsg);
+    }
+    return careerFunctions;
+}
+
+# Retrieve compensation data of a given filter.
+#
+# + filter - Filter to retrieve the compensation data
+# + return - Compensation data or error
+public isolated function getCompensation(string filter) returns types:CompensationEmail|error {
+
+    CompensationEmail|error compensationRecord = databaseClient->queryRow(getCompensationQuery(filter));
+    if compensationRecord is error {
+        if compensationRecord is sql:NoRowsError {
+            return error(string `No matching compensation data found for ${filter}!`);
+        }
+        string customError = string `Error occurred while retrieving the compensation data for ${filter}!`;
+        log:printError(customError, compensationRecord);
+        return error(customError);
+    }
+
+    types:Compensation[]|error compensation = compensationRecord.keyValuePairs.fromJsonStringWithType();
+    if compensation is error {
+        string customError = string `Error occurred while converting compensation data for ${filter}!`;
+        log:printError(customError, compensation);
+        return error(customError);
+    }
+    if compensation.length() == 0 {
+        return error(string `No matching compensation data found for ${filter}`);
+    }
+
+    return {emailTemplate: compensationRecord.emailTemplate, compensation};
+}
+
+# Retrieve companies under from HRIS.
+#
+# + filter - Criteria to filter the data  
+# + 'limit - Number of records to retrieve  
+# + offset - Number of records to offset
+# + return - List of companies
+public isolated function getCompanies(CompanyFilter filter, int 'limit, int offset)
+    returns Company[]|error {
+
+    stream<DBCompany, sql:Error?> resultStream = databaseClient->query(getCompaniesQuery(filter, 'limit, offset));
+    Company[] companies = [];
+    error? iterateError = from DBCompany company in resultStream
+        do {
+            Office[]|error offices = company.offices.fromJsonStringWithType();
+            if offices is error {
+                string errorMsg = string `An error occurred when retrieving offices data of ${company.company} !`;
+                log:printError(errorMsg, offices);
+                return error(errorMsg);
```

Comment:
```suggestion
                return error(errorMsg, offices);
```

Reply:
Since we logged the error withthe stack trace in the resource function, we don't need to do it here. We can just pass the original error with the message.